### PR TITLE
Update TF-M to v1.3.0

### PR DESCRIPTION
**Pull request by LDong-Arm**
Preceding PR: https://github.com/ARMmbed/mbed-os/pull/14582

This PR integrates TF-M v1.3.0 with Mbed OS.

* Starting from this version, we use unmodified TF-M from [git.trustedfirmware.org](https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/) rather than [the ARMmbed fork](https://github.com/ARMmbed/trusted-firmware-m). The fetching is changed accordingly:
  * Use trusted-firmware-m's own version control
  * Clean up version tracking of trusted-firmware-m
  * Build TF-M v1.3.0 release
* Changes related to the build itself:
  * Update the list of files to import from TF-M
  * Disable __stdout from libplatform_ns
  * TF-M IRQ API now removed for the IPC model
  * Enable the PSA IPC test suite on Musca S1
* General improvements, not required for the integration
  * Handle switching of git remote
  * Remove unused headers
  * Travis: Use mbed-tools deploy
* To be removed after the preceding PR is merged
  * TEMPORARY: use ARMmbed/mbed-os#14582

See individual commits for details.

### CODE REVIEW

File: tfm_ns_import.yaml [hunk 1]
```diff
@@ -178,25 +190,25 @@
                 "dst": "tfm/platform/include/tfm_plat_ns.h"
             },
             {
-                "src": "../../tf-m-tests/test/test_services/tfm_secure_client_service/tfm_secure_client_service_api.h",
+                "src": "cmake_build/lib/ext/tfm_test_repo-src/test/test_services/tfm_secure_client_service/tfm_secure_client_service_api.h",
```
Comment (Patater) [inline tfm_ns_import.yaml]: Where does `cmake_build` come from? Would it be present if we use the `--cli 1` option?

File: build_tfm.py [hunk 2]
```diff
@@ -519,6 +519,23 @@ def _copy_library(source, toolchain):
...
                     logging.info("Skipping " + src_file)
 
+                if dst_base == "libplatform_ns.ar":
+                    # TF-M redirects output to serial by declaring its own `FILE __stdout`
+                    # and disables the toolchain's default version of this symbol using
+                    # the flag `-nostdlib`. But stdlib is enabled and required by Mbed OS,
+                    # so we need to disable the one from TF-M's libplatform_ns to avoid
+                    # symbol duplication.
+                    cmd = [
+                        "fromelf",
```
Comment (Patater) [inline build_tfm.py:528]: Note to myself to verify we have this dependency documented in the README

File: main.cpp [hunk 3]
```diff
@@ -60,12 +52,12 @@ int main(void)
...
 #endif
 
-    // Disable deep sleep
-    sleep_manager_lock_deep_sleep();
-
     tfm_log_printf("Starting TF-M PSA API tests\r\n");
 
-    return val_entry();
+    uint32_t retval = val_entry();
+    TEST_ASSERT_EQUAL_UINT32(0, retval);
```
Comment (Patater) [inline main.cpp:57]: Moving to TEST_ASSERT here on the return value is a change that deserves its own commit and justification. It's a good change, just doesn't fit the `TF-M IRQ API now removed for the IPC model` commit's purpose.

File: build_tfm.py [hunk 4]
```diff
@@ -567,6 +567,9 @@ def _build_target(tgt, cmake_build_dir, args):
 
         _copy_tfm_ns_files(cmake_build_dir, tgt[0])
 
+    if args.commit:
+        _commit_changes(MBED_TF_M_PATH)
```
Comment (Patater) [inline build_tfm.py:571]: What commit message does this make? Did we need to add this here? Should we remove the other call to `_commit_changes(MBED_TF_M_PATH)` to avoid two commits when one combined one is best?

File: build_tfm.py [hunk 5]
```diff
@@ -71,7 +71,7 @@ def _clone_tfm_repo(commit):
     Clone TF-M git repos and it's dependencies
     :param commit: If True then commit VERSION.txt
     """
-    check_and_clone_repo("trusted-firmware-m", "latest-tfm", TF_M_BUILD_DIR)
+    check_and_clone_repo("trusted-firmware-m", "released-tfm", TF_M_BUILD_DIR)
```
Comment (Patater) [inline build_tfm.py:74]: Commit message "Build TF-M v1.3.0 release" seems like a misnomer. It's more like "Build the most recently released version of TF-M"

File: .travis.yml [hunk 6]
```diff
@@ -72,10 +72,9 @@ matrix:
         - sudo ln -s $(which ccache) bin/arm-none-eabi-gcc
         - sudo ln -s $(which ccache) bin/arm-none-eabi-g++
         - export PATH="$(pwd)/bin:$PATH"
-        # Fetch mbed-os: We use manual clone, with depth=1 and --single-branch to save time.
-        - git clone --depth=1 --single-branch https://github.com/ARMmbed/mbed-os.git
         # Install Mbed CLI and dependencies
         - pip install --upgrade mbed-tools
+        - mbed-tools deploy
```
Comment (Patater) [inline .travis.yml]: Should we make this change for CircleCI as well?

File: .travis.yml [hunk 7]
```diff
@@ -74,7 +74,9 @@ matrix:
         - export PATH="$(pwd)/bin:$PATH"
         # Install Mbed CLI and dependencies
         - pip install --upgrade mbed-tools
-        - mbed-tools deploy
+        # TEMPORARY: Use ARMmbed/mbed-os/pull#14582
+        # - mbed-tools deploy
+        - git clone https://github.com/LDong-Arm/mbed-os -b TF-Mv1.3.0_update --depth=1
```
Comment (Patater) [inline .travis.yml:79]: Doesn't this branch get picked up from mbed-os.lib?

**Review by Patater — CHANGES_REQUESTED** [review]

**Comment by Patater** [comment]
Rebase on latest master

File: psa_builder.py [hunk 8]
```diff
@@ -237,7 +236,32 @@ def check_and_clone_repo(name, deps, dir):
...
+                tfm_dir,
+                "remote",
+                "add",
+                deps,
+                dependencies[deps].get(name)[0],
+            ]
+            ret = run_cmd_and_return(cmd)
+            if ret != 0:
+                logging.critical("Failed to add remote %s", deps)
+                sys.exit(1)
```
Comment (rwalton-arm) [inline psa_builder.py:262]: This is kind of preexisting as you're just following existing patterns... Shouldn't we just wrap `_main` with a `sys.exit` and come up with a proper exception handling strategy? If I ever want to call this from some other code it could possibly exit the interpreter without giving me any chance to handle the error, which isn't very considerate.

File: psa_builder.py [hunk 9]
```diff
@@ -237,7 +236,32 @@ def check_and_clone_repo(name, deps, dir):
...
+                "git",
+                "-C",
+                tfm_dir,
+                "remote",
+                "add",
+                deps,
+                dependencies[deps].get(name)[0],
+            ]
+            ret = run_cmd_and_return(cmd)
+            if ret != 0:
```
Comment (rwalton-arm) [inline psa_builder.py:260]: We could return a boolean or define a `Enum` with possible return codes here. In Python you want to probably use a higher level construct than an int to signify a failure. We could even use exceptions... 

How do we know just from the return code what the actual error was? We assume it's a failure to add a remote or a failure to find a remote. Could we make this more explicit by either raising a specific exception or returning a more specific error code?

File: build_tfm.py [hunk 10]
```diff
@@ -18,7 +18,7 @@
 """
 
 import os
-from os.path import join, abspath, dirname, isdir, relpath, split
+from os.path import join, abspath, dirname, exists, isdir, relpath, split
```
Comment (rwalton-arm) [inline build_tfm.py:21]: The generally recommended way to import/use functions from `os.path` is to `import os` and call the functions with the qualified namespace e.g `os.path.join` -- this makes it obvious to experienced Python programmers where these functions have come from, and removes ambiguity from code that calls `os.path.join` and `str.join`.

File: build_tfm.py [hunk 11]
```diff
@@ -519,6 +519,23 @@ def _copy_library(source, toolchain):
...
+                    cmd = [
+                        "fromelf",
+                        "--elf",
+                        "--localize",
+                        "__stdout",
+                        dst_file,
+                        "-o",
+                        dst_file,
+                    ]
+                    run_cmd_and_return(cmd)
```
Comment (rwalton-arm) [inline build_tfm.py:536]: Do we not care about errors here?

**Review by rwalton-arm — DISMISSED** [review]

File: tfm_ns_import.yaml (hunk 1 above)
Comment (Patater) [reply tfm_ns_import.yaml]: OK, this is from `build_tfm.py:585`:
```
585:    cmake_build_dir = join(TF_M_BUILD_DIR, "trusted-firmware-m", "cmake_build")
```

File: build_tfm.py (hunk 2 above)
Comment (Patater) [reply build_tfm.py:528]: This is part of the `ARMCLANG` toolchain, and within an ifdef for `ARMCLANG`.

File: build_tfm.py (hunk 4 above)
Comment (Patater) [reply build_tfm.py:571]: There are indeed two commits. Instead of a unified "Upgrade TF-M for such and such platform", we have a point in history where the binary and the supporting files are out of sync. Also, there are `"` in the commit messages themselves...

```
ec76f07b91e7 (HEAD) "Updated directory platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST"
ddb94b8ee8ae "Updated secure binaries for [('ARM_MUSCA_B1', 'GNUARM')]"
```

File: .travis.yml (hunk 6 above)
Comment (Patater) [reply .travis.yml]: I did so. This means Circle CI will use what's in `mbed-os.lib` rather than the latest `master` branch of Mbed OS, so it's a change of behavior (though maybe a welcome one).

File: .travis.yml (hunk 7 above)
Comment (Patater) [reply .travis.yml:79]: It does. Removed this so we only update mbed-os.lib.

File: psa_builder.py (hunk 8 above)
Comment (Patater) [reply psa_builder.py:262]: Filed https://github.com/ARMmbed/mbed-os-tf-m-regression-tests/issues/104 to address the whole codebase

File: psa_builder.py (hunk 9 above)
Comment (Patater) [reply psa_builder.py:260]: Yes, we should use exceptions more than we do. Here, we don't care what the error is very much. If anything failed, we explain that we failed what we attempted to do. This can be unhelpful as we aren't explaining _why_ we failed.

**Comment by Patater** [comment]
Updated to address review comments

File: psa_builder.py (hunk 9 above)
Comment (rwalton-arm) [reply psa_builder.py:259]: The error code and exception stuff is probably covered by #104. I'm assuming we can't explain why we failed until the error handling is reworked so this function produces more specific return codes or exceptions.

**Comment by Patater** [comment]
Fixed black style in each commit.

**Comment by Patater** [comment]
Fixed black style in each commit, using a newer black version.

**Comment by Patater** [comment]
Updated with black again, using `black . -l 79` to match CI.

**Comment by Patater** [comment]
The preceding Mbed OS PR has been merged now. Remove the TEMPORARY commit that points to Lingkai's Mbed OS branch.

**Review by Patater — APPROVED** [review]
